include/zephyr/fs/zms.h

```diff
@@ -0,0 +1,215 @@
+ * to be written. When a rewrite of the same data already stored is attempted, nothing is written
+ * to flash, thus 0 is returned. On error, returns negative value of errno.h defined error codes.
+ */
+ssize_t zms_write(struct zms_fs *fs, uint32_t id, const void *data, size_t len);
+
+/**
+ * @brief Delete an entry from the file system
+ *
+ * @param fs Pointer to file system
+ * @param id Id of the entry to be deleted
+ * @retval 0 Success
+ * @retval -ERRNO errno code if error
+ */
+int zms_delete(struct zms_fs *fs, uint32_t id);
+
+/**
+ * @brief Read an entry from the file system.
+ *
+ * @param fs Pointer to file system
+ * @param id Id of the entry to be read
+ * @param data Pointer to data buffer
+ * @param len Number of bytes to be read (or size of the allocated read buffer)
+ *
+ * @return Number of bytes read. On success, it will be equal to the number of bytes requested
+ * to be read. When the return value is less than the number of bytes requested to read this
+ * indicates that ATE contain less data than requested. On error, returns negative value of
+ * errno.h defined error codes.
+ */
+ssize_t zms_read(struct zms_fs *fs, uint32_t id, void *data, size_t len);
+
+/**
+ * @brief Read a history entry from the file system.
+ *
+ * @param fs Pointer to file system
+ * @param id Id of the entry to be read
+ * @param data Pointer to data buffer
+ * @param len Number of bytes to be read
+ * @param cnt History counter: 0: latest entry, 1: one before latest ...
+ *
+ * @return Number of bytes read. On success, it will be equal to the number of bytes requested
+ * to be read. When the return value is larger than the number of bytes requested to read this
+ * indicates not all bytes were read, and more data is available. On error, returns negative
+ * value of errno.h defined error codes.
+ */
+ssize_t zms_read_hist(struct zms_fs *fs, uint32_t id, void *data, size_t len, uint32_t cnt);
+
+/**
+ * @brief Gets the data size that is stored in an entry with a given id
+ *
+ * @param fs Pointer to file system
+ * @param id Id of the entry that we want to get its data length
+ *
+ * @return Data length contained in the ATE. On success, it will be equal to the number of bytes
+ * in the ATE. On error, returns negative value of errno.h defined error codes.
+ */
+ssize_t zms_get_data_length(struct zms_fs *fs, uint32_t id);
+/**
+ * @brief Calculate the available free space in the file system.
+ *
+ * @param fs Pointer to file system
+ *
+ * @return Number of bytes free. On success, it will be equal to the number of bytes that can
+ * still be written to the file system.
+ * Calculating the free space is a time consuming operation, especially on spi flash.
+ * On error, returns negative value of errno.h defined error codes.
+ */
+ssize_t zms_calc_free_space(struct zms_fs *fs);
+
+/**
+ * @brief Tell how many contiguous free space remains in the currently active ZMS sector.
+ *
+ * @param fs Pointer to the file system.
```

Comment:
(minor) I like this version more. But it's also inconsistent with the functions declared above. Could be nice to have them all aligned.